SOLR-15423 JWTAuthPlugin support for custom truststore

janhoy opened the pull request:
https://issues.apache.org/jira/browse/SOLR-15423

janhoy commented:
Feedback welcome. I could not think of a good way to actually test that the certificates are used (do we have a mock HTTPS server somewhere?).

janhoy commented on solr/core/src/java/org/apache/solr/security/JWTAuthPlugin.java:
```diff
@@ -155,8 +168,33 @@ public void init(Map<String, Object> pluginConfig) {
+    }
+    if (trustedCertsFile != null) {
+      try {
+          trustedCertsStream = Files.newInputStream(Paths.get(trustedCertsFile));
```
Hmm, I tried using `CoreContainer.assertPathAllowed()` but unfortunately AuthPlugins do not have access to coreContainer, so I cannot get access to it. An Auth plugin is at the container level, not the core level, so what is the equivalent of `SolrCoreAware` for such plugins?  I.e. that it would get a reference to `CoreContainer`?

madrob replied:
CoreContainer:517 attempts to create an authentication plugin using a constructor with the CoreContainer as a parameter first, then falls back to a zero-arg constructor. I think you can just add this constructor and hold a reference to the container.

janhoy replied:
Thanks a lot - it actually worked!!

Solr should probably replace its dependency injection system at some point :)

madrob commented on solr/core/src/java/org/apache/solr/security/JWTAuthPlugin.java:
```diff
@@ -199,6 +237,28 @@ public void init(Map<String, Object> pluginConfig) {
+   * @param trustedCertsStream input stream with the contents of either PEM (plaintext) or DER (binary) certs
+   * @return collection of found certificates, else throws exception
+   */
+  static Collection<X509Certificate> parsePemToX509Certs(InputStream trustedCertsStream) {
```
Maybe this belongs in CryptoKeys? Or some other util class?

janhoy replied:
Good idea - moving to CryptoKeys and generalizing a bit

madrob commented on solr/core/src/java/org/apache/solr/security/JWTIssuerConfig.java:
```diff
@@ -335,33 +346,49 @@ public boolean isValid() {
+      if (trustedCerts != null) {
+        Get getWithCustomTrust = new Get();
+        getWithCustomTrust.setTrustedCertificates(trustedCerts);
+        log.info("Trusting custom SSL certificate(s) for the IdP");
```
How often does this happen? Is it per user request? Per user session? Might be too verbose.

janhoy replied:
I moved ths to the first config parsing in JwtAuthPlugin instead. It will print on every re-initialization / change of security.json.

madrob commented on solr/solr-ref-guide/src/jwt-authentication-plugin.adoc:
```diff
@@ -54,6 +54,8 @@ rolesClaim           ; What claim id to pull user roles from. The claim must the
 claimsMatch          ; JSON object of claims (key) that must match a regular expression (value). Example: `{ "foo" : "A|B" }` will require the `foo` claim to be either "A" or "B". ;
 adminUiScope         ; Define what scope is requested when logging in from Admin UI ; If not defined, the first scope from `scope` parameter is used
 redirectUris         ; Valid location(s) for redirect after external authentication. Takes a string or array of strings. Must be the base URL of Solr, e.g., https://solr1.example.com:8983/solr/ and must match the list of redirect URIs registered with the Identity Provider beforehand. ; Defaults to empty list, i.e., any node is assumed to be a valid redirect target.
+trustedCerts         ; One or more X.509 SSL certificates in plaintext PEM or PKCS#7 formats, that should be trusted when talking to IdPs. Newlines must be replaced with `\n`. ; Defaults to Java truststore
```
I'd like to see more explanation when somebody would need this, similar to what was described in the JIRA, e.g. explain that Jetty's TrustStore is ignored if running with HTTP with an SSL-terminating proxy/LB/ingress in front.

janhoy replied:
I added a separate paragraph explaining more, and linked to it from the table.

madrob commented on solr/core/src/test/org/apache/solr/security/JWTAuthPluginTest.java:
```diff
@@ -477,4 +484,51 @@ public void initWithToplevelAndIssuersCombined() {
+    authConf.put("jwksUrl", "https://127.0.0.1:9999/foo.jwk");
+    authConf.put("trustedCertsFile", TEST_PATH().resolve("security").resolve("jwt_idpCert.pem").toString());
+    plugin = new JWTAuthPlugin();
+    plugin.init(authConf);
```
Anything we can do to assert that the generated factory has the correct trust store?

janhoy replied:
I added a test to check that there are 2 X509 certs in IssuerConfig. But still need to test somehow that the certs are actually used on outbound traffic to WellKnown endpoint as well as JWKS endpoint.

janhoy commented:
Committed latest changes. Please resolve outstanding conversations if you're happy. Still would like some integration test to check the actual certificate usage. Think I'll try to spin up a MockIdp listening on some SSL socket...

janhoy commented:
I'm trying to integrate https://github.com/navikt/mock-oauth2-server into the tests (actually open-sourced by a Norwegian government entity). Quite simple, you initialize a `MockOAuth2Server` and then it listens on a separate port and behaves like an Oauth server. I'm struggling to make it talk https though, so that we can finally assert that the new config options actually works.

If this ends up working, we can also add several other end-to-end integration tests for JWT auth (in separate issues).

janhoy commented:
I have now addressed my two points
* Create a negative test with a different cert, to see it failing
* Remove need for BouncyCastle by importing pcs12 file directly in test

I have also added license and sha files for all new test dependencies and made precommit pass.
I think this is now ready for review.
It's a large PR, but most files are really SHA, LICENSE and NOTICE files.

janhoy commented on solr/core/build.gradle:
```diff
@@ -149,5 +149,12 @@ dependencies {
   testImplementation('org.mockito:mockito-core', {
     exclude group: "net.bytebuddy", module: "byte-buddy-agent"
   })
+  testImplementation('no.nav.security:mock-oauth2-server', {
```
I add this as a testImpl dependency, but still I'm forced to add LICENSE and NOTICE files in the `licenses` folder for this and all its dependencies. Why is that necessary, when we do not distribute these jars, they are only used during tests?
There are the new (test) deps:
```
+--- no.nav.security:mock-oauth2-server -> 0.3.3
|    +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.32
|    |    +--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.32
|    |    \--- org.jetbrains:annotations:13.0
|    +--- org.jetbrains.kotlin:kotlin-reflect:1.4.32
|    |    \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.32 (*)
|    +--- com.fasterxml.jackson.core:jackson-databind:2.12.2 (*)
|    +--- io.github.microutils:kotlin-logging:2.0.6
|    |    \--- io.github.microutils:kotlin-logging-jvm:2.0.6
|    |         +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.31 -> 1.4.32 (*)
|    |         \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.31 -> 1.4.32
|    +--- com.fasterxml.jackson.module:jackson-module-kotlin:2.12.2
|    |    +--- com.fasterxml.jackson.core:jackson-databind:2.12.2 (*)
|    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.12.2 (*)
|    |    +--- org.jetbrains.kotlin:kotlin-reflect:1.4.21 -> 1.4.32 (*)
|    |    \--- com.fasterxml.jackson:jackson-bom:2.12.2 (*)
|    +--- org.freemarker:freemarker:2.3.31
|    +--- com.squareup.okhttp3:mockwebserver:4.9.1
|    |    +--- com.squareup.okhttp3:okhttp:4.9.1
|    |    |    +--- com.squareup.okio:okio:2.8.0
|    |    |    |    +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.0 -> 1.4.32 (*)
|    |    |    |    \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.0 -> 1.4.32
|    |    |    \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 -> 1.4.32 (*)
|    |    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.10
|    |         +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 -> 1.4.32 (*)
|    |         \--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.4.10
|    |              \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 -> 1.4.32 (*)
|    \--- com.nimbusds:oauth2-oidc-sdk:9.2.3
|         +--- com.github.stephenc.jcip:jcip-annotations:1.0-1
|         +--- com.nimbusds:content-type:2.1
|         +--- net.minidev:json-smart:[1.3.1,2.3] -> 2.3
|         |    \--- net.minidev:accessors-smart:1.2
|         |         \--- org.ow2.asm:asm:5.0.4 -> 7.2
|         +--- com.nimbusds:lang-tag:1.4.4
|         \--- com.nimbusds:nimbus-jose-jwt:9.6.1
|              \--- com.github.stephenc.jcip:jcip-annotations:1.0-1
```

janhoy replied:
Anyone knows about this?

janhoy replied:
See https://issues.apache.org/jira/browse/SOLR-15465 for a separate discussion of this topic, leaving it in for now.

madrob commented on solr/core/src/java/org/apache/solr/security/JWTVerificationkeyResolver.java:
```diff
@@ -110,7 +112,13 @@ public Key resolveKey(JsonWebSignature jws, List<JsonWebStructure> nestingContex
+          try {
+            jsonWebKeys.addAll(hjwks.getJsonWebKeys());
+          } catch (SSLHandshakeException e) {
+            log.warn("Failed to talk SSL with {}, do you have the correct SSL certificate configured?", hjwks.getLocation());
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+                "SSL error when validating auth token. Make sure correct certificates are configured.", e);
```
Generally not a fan of both logging and throwing, as that likely means that we log twice.

janhoy replied:
Agree, now only throws exception

madrob commented on solr/solr-ref-guide/src/jwt-authentication-plugin.adoc:
```diff
@@ -158,12 +161,18 @@ Let's comment on this config:
+=== Trusting the IdP server
+All communication with issuer server (IdP) is done over HTTPS. If the issuer uses a self signed SSL certificate, the connection will fail since the plugin enforces verification of SSL certificates on outbound traffic. By default, the root certificates present in Java's TrustStore (default or custom provided) are consulted. If you do not want to place the IdP certificates in Java's TrustStore, you ...
+
+Please configure either `trustedCerts` or `trustedCertsFile`, not both.
```
Stronger language here, indicating that it will fail if both are configured.

janhoy replied:
Done.

madrob commented on solr/solr-ref-guide/src/jwt-authentication-plugin.adoc:
```diff
@@ -158,12 +161,18 @@ Let's comment on this config:
 security check that Solr performs. To support this you can set the environment variable `solr.auth.jwt.allowOutboundHttp=true`.
 
+=== Trusting the IdP server
+All communication with issuer server (IdP) is done over HTTPS. If the issuer uses a self signed SSL certificate, the connection will fail since the plugin enforces verification of SSL certificates on outbound traffic. By default, the root certificates present in Java's TrustStore (default or custom provided) are consulted. If you do not want to place the IdP certificates in Java's TrustStore, you ...
```
I like this comment.

madrob commented on solr/core/src/java/org/apache/solr/security/JWTAuthPlugin.java:
```diff
@@ -155,8 +175,38 @@ public void init(Map<String, Object> pluginConfig) {
+      }
+    }
+    if (trustedCerts != null) {
+      trustedCertsStream = IOUtils.toInputStream(trustedCerts, StandardCharsets.UTF_8);
```
I'm amazed that this doesn't exist in the JDK. We have StringStream in ContentStreamBase that is similar, maybe we can create something for the common need.

janhoy replied:
Is there a problem using IOUtils for this?

madrob commented on solr/core/src/test/org/apache/solr/security/JWTAuthPluginIntegrationTest.java:
```diff
@@ -70,54 +89,43 @@
+  @Override
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
```
if all it does it call super, then we don't need this method.

janhoy replied:
Agree, removed it.

madrob commented on solr/core/src/test/org/apache/solr/security/JWTAuthPluginIntegrationTest.java:
```diff
@@ -127,13 +135,40 @@ public void tearDown() throws Exception {
-  public void infoRequestWithoutToken() throws Exception {
-    get(baseUrl + "/admin/info/system", null);
+  @Test
+  public void extractCertificateFromPem() throws IOException {
```
Can we have a test where the pen file or directly supplied string is improperly formatted? I'd like to see what kind of error message users can expect in that case (and maybe we should document that, because I suspect it will be cryptic otherwise)

janhoy replied:
I added two tests to JWTAuthPluginTest, verifying that init() fails. No need for integration test of all these? 

If you load a `security.json` to ZK with bogous `trustedCertsFile` or `trustedCerts`, the JWT plugin will throw and exception and auth will not be enabled. Here are log messages you can expect to see prior to exception:

> Reading trustedCerts from file {}

or

> Reading trustedCerts PEM from configuration string

In the exception stacktrace itself  you'll see either

> Wrong type of certificates. Must be DER or PEM format

or

> Failed loading certificate(s) from input stream

madrob reviewed (commented):
Overall looks good, few minor comments. I did not run tests on this.

janhoy commented:
> BUILD SUCCESSFUL in 36m 16s
I'll merge this now
